<a href="https://colab.research.google.com/github/ashzmi/test/blob/master/Ecommerce_Hackathon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Libraries

In [193]:
#analysis the data and wrangling
import pandas as pd
import numpy as np
import random as rnd

# Importing E-Commerce dataset with csv format

In [194]:
session_df = pd.read_csv('session.csv')
session_df.head()


,session_id,customer_id,start_time,device,source,country
0,1,12360,2021-12-27T00:01:36,mobile,email,DE
1,2,13917,2025-01-31T21:29:42,desktop,organic,PL
2,3,1022,2024-02-19T00:52:50,tablet,organic,FR
3,4,2882,2024-08-04T19:54:31,mobile,direct,GB
4,5,1286,2022-06-28T13:58:08,desktop,email,ES


In [195]:
customers_df = pd.read_csv('customers.csv')

print(customers_df.head())

   customer_id              name                       email country  age  \
0            1  Jennifer Salinas      nicholas59@example.org      JP   71   
1            2     Phillip Ramos  christinarubio@example.com      IN   26   
2            3       Dawn Fowler       jessica03@example.org      BR   21   
3            4      Mario Butler         paula27@example.org      FR   63   
4            5       Amber Brown         kevin85@example.net      BR   19   

  signup_date  marketing_opt_in  
0  2020-09-04              True  
1  2020-04-05             False  
2  2023-08-31              True  
3  2022-06-30              True  
4  2022-07-22              True  


In [196]:
events_df = pd.read_csv('events.csv')

print(events_df.head())

   event_id  session_id            timestamp   event_type  product_id  qty  \
0         1           1  2021-12-27T00:08:36    page_view        93.0  NaN   
1         2           1  2021-12-27T00:16:36    page_view      1005.0  NaN   
2         3           1  2021-12-27T00:18:01  add_to_cart      1005.0  1.0   
3         4           1  2021-12-27T00:45:36    page_view       918.0  NaN   
4         5           1  2021-12-27T01:03:36    page_view       946.0  NaN   

   cart_size payment  discount_pct  amount_usd  
0        NaN     NaN           NaN         NaN  
1        NaN     NaN           NaN         NaN  
2        NaN     NaN           NaN         NaN  
3        NaN     NaN           NaN         NaN  
4        NaN     NaN           NaN         NaN  


In [197]:
orderItems_df = pd.read_csv('order_items.csv')

print(orderItems_df.head())

   order_id  product_id  unit_price_usd  quantity  line_total_usd
0         1         226          107.15         1          107.15
1         2         771          116.17         1          116.17
2         3         415           94.49         1           94.49
3         3          24           42.86         1           42.86
4         4        1157           32.18         1           32.18


In [198]:
orders_df = pd.read_csv('orders.csv')

print(orders_df.head())

   order_id  customer_id           order_time payment_method  discount_pct  \
0         1        13917  2025-01-31T23:07:42           card            20   
1         2         1022  2024-02-19T01:17:50           card             0   
2         3         6145  2024-12-04T20:24:13           card             0   
3         4         3152  2024-07-17T08:50:47           card            15   
4         5        12378  2020-08-21T16:54:16           card             0   

   subtotal_usd  total_usd country   device   source  
0        107.15      85.72      PL  desktop  organic  
1        116.17     116.17      FR   tablet  organic  
2        137.35     137.35      US   mobile  organic  
3         32.18      27.35      BR   mobile    email  
4        238.09     238.09      NL  desktop     paid  


In [199]:
products_df = pd.read_csv('products.csv')

print(products_df.head())

   product_id     category                      name  price_usd  cost_usd  \
0           1  Electronics        SSD MediumBlue 149     570.28    352.69   
1           2  Electronics     Keyboard DeepPink 696     498.13    263.13   
2           3  Electronics     Headphones Orchid 188     548.53    309.60   
3           4  Electronics  Smartwatch BurlyWood 664     268.36    153.56   
4           5  Electronics   Smartwatch Cornsilk 328      63.69     42.65   

   margin_usd  
0      217.59  
1      235.00  
2      238.93  
3      114.80  
4       21.04  


In [200]:
reviews_df = pd.read_csv('reviews.csv')

print(reviews_df.head())

   review_id  order_id  product_id  rating  \
0          1         4        1157       2   
1          2         7         405       4   
2          3         7         487       5   
3          4         7         442       5   
4          5         7         348       4   

                            review_text          review_time  
0                    Not great quality.  2022-02-01T00:00:00  
1                 Good value for money!  2020-03-07T00:00:00  
2  Excellent product, highly recommend!  2024-12-26T00:00:00  
3  Excellent product, highly recommend!  2021-03-18T00:00:00  
4                 Good value for money!  2025-09-30T00:00:00  


# Analyze, Cleaning and Transform Datasets

# Analyse session_df Dataset

In [201]:
session_df.head(10)

,session_id,customer_id,start_time,device,source,country
0,1,12360,2021-12-27T00:01:36,mobile,email,DE
1,2,13917,2025-01-31T21:29:42,desktop,organic,PL
2,3,1022,2024-02-19T00:52:50,tablet,organic,FR
3,4,2882,2024-08-04T19:54:31,mobile,direct,GB
4,5,1286,2022-06-28T13:58:08,desktop,email,ES
5,6,2319,2020-07-22T20:10:01,mobile,paid,SG
6,7,7050,2023-05-07T14:35:49,mobile,email,GB
7,8,5687,2020-07-10T10:35:54,mobile,email,DE
8,9,6476,2024-02-22T16:57:44,mobile,direct,PL
9,10,6145,2024-12-04T18:53:13,mobile,organic,US


In [202]:
session_df.shape

(120000, 6)

In [203]:
session_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 6 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   session_id   120000 non-null  int64 
 1   customer_id  120000 non-null  int64 
 2   start_time   120000 non-null  object
 3   device       120000 non-null  object
 4   source       120000 non-null  object
 5   country      120000 non-null  object
dtypes: int64(2), object(4)
memory usage: 5.5+ MB


In [204]:
print (session_df.columns.values)

['session_id' 'customer_id' 'start_time' 'device' 'source' 'country']


In [205]:
#showing the datatypes of the column
session_df.dtypes

,0
session_id,int64
customer_id,int64
start_time,object
device,object
source,object
country,object


In [206]:
print(session_df.count())

session_id     120000
customer_id    120000
start_time     120000
device         120000
source         120000
country        120000
dtype: int64


In [207]:
#describe the dataframe with showing the count, mean, standard deviation, min, percentiles, and max.
session_df.describe()

,session_id,customer_id
count,120000.000000,120000.000000
mean,60000.500000,10020.814533
std,34641.160489,5790.838535
min,1.000000,1.000000
25%,30000.750000,5003.000000
50%,60000.500000,10023.000000
75%,90000.250000,15052.000000
max,120000.000000,20000.000000


Check Null value

In [208]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
session_df.isna().sum()

Number of NaN values for each columns: 


,0
session_id,0
customer_id,0
start_time,0
device,0
source,0
country,0


In [209]:
#check duplicate
session_df.duplicated().sum()

np.int64(0)

In [210]:
session_df.head()

,session_id,customer_id,start_time,device,source,country
0,1,12360,2021-12-27T00:01:36,mobile,email,DE
1,2,13917,2025-01-31T21:29:42,desktop,organic,PL
2,3,1022,2024-02-19T00:52:50,tablet,organic,FR
3,4,2882,2024-08-04T19:54:31,mobile,direct,GB
4,5,1286,2022-06-28T13:58:08,desktop,email,ES


In [211]:
session_df['start_time'] = pd.to_datetime(session_df['start_time'])

In [212]:
session_df['session_date'] = session_df['start_time'].dt.date

In [213]:
session_df['session_time'] = session_df['start_time'].dt.time

In [214]:
session_df[['start_time', 'session_date', 'session_time']].head()

,start_time,session_date,session_time
0,2021-12-27 00:01:36,2021-12-27,00:01:36
1,2025-01-31 21:29:42,2025-01-31,21:29:42
2,2024-02-19 00:52:50,2024-02-19,00:52:50
3,2024-08-04 19:54:31,2024-08-04,19:54:31
4,2022-06-28 13:58:08,2022-06-28,13:58:08


In [215]:
session_df.head()

,session_id,customer_id,start_time,device,source,country,session_date,session_time
0,1,12360,2021-12-27 00:01:36,mobile,email,DE,2021-12-27,00:01:36
1,2,13917,2025-01-31 21:29:42,desktop,organic,PL,2025-01-31,21:29:42
2,3,1022,2024-02-19 00:52:50,tablet,organic,FR,2024-02-19,00:52:50
3,4,2882,2024-08-04 19:54:31,mobile,direct,GB,2024-08-04,19:54:31
4,5,1286,2022-06-28 13:58:08,desktop,email,ES,2022-06-28,13:58:08


In [216]:
session_df=session_df.drop (columns=['start_time'])

In [217]:
session_df.head()

,session_id,customer_id,device,source,country,session_date,session_time
0,1,12360,mobile,email,DE,2021-12-27,00:01:36
1,2,13917,desktop,organic,PL,2025-01-31,21:29:42
2,3,1022,tablet,organic,FR,2024-02-19,00:52:50
3,4,2882,mobile,direct,GB,2024-08-04,19:54:31
4,5,1286,desktop,email,ES,2022-06-28,13:58:08


Change country shortform to full name

In [218]:
#list of country in session_df
sorted(session_df['country'].unique())

['AE',
 'AU',
 'BR',
 'CA',
 'DE',
 'ES',
 'FR',
 'GB',
 'IN',
 'JP',
 'MX',
 'NL',
 'PL',
 'SE',
 'SG',
 'US',
 'ZA']

In [219]:
country_map = {
    'AE': 'United Arab Emirates',
    'AU': 'Australia',
    'BR': 'Brazil',
    'CA': 'Canada',
    'DE': 'Germany',
    'ES': 'Spain',
    'FR': 'France',
    'GB': 'United Kingdom',
    'IN': 'India',
    'JP': 'Japan',
    'MX': 'Mexico',
    'NL': 'Netherlands',
    'PL': 'Poland',
    'SE': 'Sweden',
    'SG': 'Singapore',
    'US': 'United States',
    'ZA': 'South Africa'
}

session_df['country'] = session_df['country'].map(country_map)

In [220]:
#list of updated country in session_df
sorted(session_df['country'].unique())

['Australia',
 'Brazil',
 'Canada',
 'France',
 'Germany',
 'India',
 'Japan',
 'Mexico',
 'Netherlands',
 'Poland',
 'Singapore',
 'South Africa',
 'Spain',
 'Sweden',
 'United Arab Emirates',
 'United Kingdom',
 'United States']

In [221]:
session_df.head()

,session_id,customer_id,device,source,country,session_date,session_time
0,1,12360,mobile,email,Germany,2021-12-27,00:01:36
1,2,13917,desktop,organic,Poland,2025-01-31,21:29:42
2,3,1022,tablet,organic,France,2024-02-19,00:52:50
3,4,2882,mobile,direct,United Kingdom,2024-08-04,19:54:31
4,5,1286,desktop,email,Spain,2022-06-28,13:58:08


In [222]:
#recheck datatype
session_df.dtypes

,0
session_id,int64
customer_id,int64
device,object
source,object
country,object
session_date,object
session_time,object


In [223]:
#recheck the number of null values in the dataframe
print("Number of NaN values for each columns: ")
session_df.isna().sum()

Number of NaN values for each columns: 


,0
session_id,0
customer_id,0
device,0
source,0
country,0
session_date,0
session_time,0


In [224]:
#recheck duplicate
session_df.duplicated().sum()

np.int64(0)

Validate categorical column in session_df

In [225]:
session_df['session_id'].nunique()

120000

In [226]:
session_df['device'].value_counts()

,count
device,
mobile,65942
desktop,45547
tablet,8511


In [227]:
session_df['source'].value_counts()

,count
source,
organic,40776
direct,29861
paid,14465
social,14389
email,10949
referral,9560


In [228]:
session_df['country'].value_counts()

,count
country,
United States,21913
India,9567
United Kingdom,9451
Brazil,8573
Germany,8332
France,7811
Mexico,7286
Australia,6443
Canada,6057


In [229]:
#Check total number of country
session_df['country'].nunique()

17

# Analyse Customer Dataset

In [230]:
customers_df.head(10)

,customer_id,name,email,country,age,signup_date,marketing_opt_in
0,1,Jennifer Salinas,nicholas59@example.org,JP,71,2020-09-04,True
1,2,Phillip Ramos,christinarubio@example.com,IN,26,2020-04-05,False
2,3,Dawn Fowler,jessica03@example.org,BR,21,2023-08-31,True
3,4,Mario Butler,paula27@example.org,FR,63,2022-06-30,True
4,5,Amber Brown,kevin85@example.net,BR,19,2022-07-22,True
5,6,Kevin Chapman,michaelpoole@example.org,US,28,2020-10-21,False
6,7,Timothy Sanchez,shannonbrown@example.org,IN,48,2025-07-24,False
7,8,Rachael Morales,tammy08@example.com,GB,33,2022-10-09,True
8,9,Laura Guzman,awright@example.org,US,32,2024-07-29,True
9,10,Alexander Lopez,brianmaxwell@example.org,JP,39,2023-01-13,True


In [231]:
customers_df.shape

(20000, 7)

In [232]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customer_id       20000 non-null  int64 
 1   name              20000 non-null  object
 2   email             20000 non-null  object
 3   country           20000 non-null  object
 4   age               20000 non-null  int64 
 5   signup_date       20000 non-null  object
 6   marketing_opt_in  20000 non-null  bool  
dtypes: bool(1), int64(2), object(4)
memory usage: 957.2+ KB


In [233]:
print(customers_df.columns.values)

['customer_id' 'name' 'email' 'country' 'age' 'signup_date'
 'marketing_opt_in']


In [234]:
print(customers_df.count())

customer_id         20000
name                20000
email               20000
country             20000
age                 20000
signup_date         20000
marketing_opt_in    20000
dtype: int64


In [235]:
#describe the dataframe with showing the count, mean, standard deviation, min, percentiles, and max.
customers_df.describe()

,customer_id,age
count,20000.000000,20000.000000
mean,10000.500000,46.492550
std,5773.647028,16.767961
min,1.000000,18.000000
25%,5000.750000,32.000000
50%,10000.500000,46.500000
75%,15000.250000,61.000000
max,20000.000000,75.000000


Check Null Value

In [236]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
customers_df.isna().sum()

Number of NaN values for each columns: 


,0
customer_id,0
name,0
email,0
country,0
age,0
signup_date,0
marketing_opt_in,0


In [237]:
#Check duplicate
customers_df.duplicated().sum()

np.int64(0)

from null and duplicate check Customer dataframe not have null value and duplicate

Validate categorical column in customers_df

In [238]:
customers_df['country'].value_counts()

,count
country,
US,3648
IN,1589
GB,1585
BR,1421
DE,1397
FR,1325
MX,1206
AU,1045
CA,1015


In [239]:
#change country name to its full name
#based on previous country_map
print(country_map)

{'AE': 'United Arab Emirates', 'AU': 'Australia', 'BR': 'Brazil', 'CA': 'Canada', 'DE': 'Germany', 'ES': 'Spain', 'FR': 'France', 'GB': 'United Kingdom', 'IN': 'India', 'JP': 'Japan', 'MX': 'Mexico', 'NL': 'Netherlands', 'PL': 'Poland', 'SE': 'Sweden', 'SG': 'Singapore', 'US': 'United States', 'ZA': 'South Africa'}


In [240]:
#change country name to full name
customers_df['country'] = customers_df['country'].map(country_map)

In [241]:
# updated country name in customers_df
customers_df['country'].value_counts()

,count
country,
United States,3648
India,1589
United Kingdom,1585
Brazil,1421
Germany,1397
France,1325
Mexico,1206
Australia,1045
Canada,1015


In [242]:
customers_df.head()

,customer_id,name,email,country,age,signup_date,marketing_opt_in
0,1,Jennifer Salinas,nicholas59@example.org,Japan,71,2020-09-04,True
1,2,Phillip Ramos,christinarubio@example.com,India,26,2020-04-05,False
2,3,Dawn Fowler,jessica03@example.org,Brazil,21,2023-08-31,True
3,4,Mario Butler,paula27@example.org,France,63,2022-06-30,True
4,5,Amber Brown,kevin85@example.net,Brazil,19,2022-07-22,True


In Marketing option column ( marketing_opt_in ) I want to change from boolean true false to Subscribe / unsubscribe

it shows that the subscribe or not the marketing

In [243]:
# Convert boolean values in 'marketing_opt_in' to more readable labels
# True  -> Subscribed (customer agreed to receive marketing)
# False -> Unsubscribed (customer did not opt in for marketing)
customers_df['marketing_opt_in'] = customers_df['marketing_opt_in'].map({
    True: 'Subscribed',
    False: 'Unsubscribed'
})

In [244]:
customers_df.head()

,customer_id,name,email,country,age,signup_date,marketing_opt_in
0,1,Jennifer Salinas,nicholas59@example.org,Japan,71,2020-09-04,Subscribed
1,2,Phillip Ramos,christinarubio@example.com,India,26,2020-04-05,Unsubscribed
2,3,Dawn Fowler,jessica03@example.org,Brazil,21,2023-08-31,Subscribed
3,4,Mario Butler,paula27@example.org,France,63,2022-06-30,Subscribed
4,5,Amber Brown,kevin85@example.net,Brazil,19,2022-07-22,Subscribed


In [245]:
customers_df['age'].value_counts().sort_index()

,count
age,
18,330
19,352
20,367
21,361
22,326
23,339
24,339
25,350
26,349


In [246]:
customers_df['age'].nunique()

58

from output above age range are 18 -75 years old

In [247]:
# Group customer age into meaningful categories for better analysis and segmentation
customers_df['age_group'] = pd.cut(
    customers_df['age'],
    bins=[0,18,25,35,50,100],
    labels=['Teen','Young Adult','Adult','Mid Age','Senior']
)

Customer ages are grouped into categories (Teen, Young Adult, Adult, Mid Age, Senior) for easier analysis and visualization.

Age ranges:

0–18 → Teen

19–25 → Young Adult

26–35 → Adult

36–50 → Mid Age

51–100 → Senior

In [248]:
customers_df.head(10)

,customer_id,name,email,country,age,signup_date,marketing_opt_in,age_group
0,1,Jennifer Salinas,nicholas59@example.org,Japan,71,2020-09-04,Subscribed,Senior
1,2,Phillip Ramos,christinarubio@example.com,India,26,2020-04-05,Unsubscribed,Adult
2,3,Dawn Fowler,jessica03@example.org,Brazil,21,2023-08-31,Subscribed,Young Adult
3,4,Mario Butler,paula27@example.org,France,63,2022-06-30,Subscribed,Senior
4,5,Amber Brown,kevin85@example.net,Brazil,19,2022-07-22,Subscribed,Young Adult
5,6,Kevin Chapman,michaelpoole@example.org,United States,28,2020-10-21,Unsubscribed,Adult
6,7,Timothy Sanchez,shannonbrown@example.org,India,48,2025-07-24,Unsubscribed,Mid Age
7,8,Rachael Morales,tammy08@example.com,United Kingdom,33,2022-10-09,Subscribed,Adult
8,9,Laura Guzman,awright@example.org,United States,32,2024-07-29,Subscribed,Adult
9,10,Alexander Lopez,brianmaxwell@example.org,Japan,39,2023-01-13,Subscribed,Mid Age


In [249]:
customers_df = customers_df[
    ['customer_id', 'name', 'email', 'country', 'age', 'age_group',
     'signup_date', 'marketing_opt_in']
]

In [250]:
print(customers_df.columns.values)

['customer_id' 'name' 'email' 'country' 'age' 'age_group' 'signup_date'
 'marketing_opt_in']


# Analyse Event Dataset

In [251]:
events_df.head(10)

,event_id,session_id,timestamp,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd
0,1,1,2021-12-27T00:08:36,page_view,93.0,NaN,NaN,NaN,NaN,NaN
1,2,1,2021-12-27T00:16:36,page_view,1005.0,NaN,NaN,NaN,NaN,NaN
2,3,1,2021-12-27T00:18:01,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN
3,4,1,2021-12-27T00:45:36,page_view,918.0,NaN,NaN,NaN,NaN,NaN
4,5,1,2021-12-27T01:03:36,page_view,946.0,NaN,NaN,NaN,NaN,NaN
5,6,1,2021-12-27T01:05:05,add_to_cart,946.0,1.0,NaN,NaN,NaN,NaN
6,7,1,2021-12-27T01:18:36,page_view,915.0,NaN,NaN,NaN,NaN,NaN
7,8,1,2021-12-27T01:37:36,page_view,931.0,NaN,NaN,NaN,NaN,NaN
8,9,1,2021-12-27T01:39:00,add_to_cart,931.0,1.0,NaN,NaN,NaN,NaN
9,10,1,2021-12-27T01:59:36,page_view,978.0,NaN,NaN,NaN,NaN,NaN


In [252]:
events_df.shape

(760958, 10)

In [253]:
events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 760958 entries, 0 to 760957
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   event_id      760958 non-null  int64  
 1   session_id    760958 non-null  int64  
 2   timestamp     760958 non-null  object 
 3   event_type    760958 non-null  object 
 4   product_id    682469 non-null  float64
 5   qty           143126 non-null  float64
 6   cart_size     44909 non-null   float64
 7   payment       33580 non-null   object 
 8   discount_pct  33580 non-null   float64
 9   amount_usd    33580 non-null   float64
dtypes: float64(5), int64(2), object(3)
memory usage: 58.1+ MB


In [254]:
events_df['timestamp'] = pd.to_datetime(events_df['timestamp'], errors='coerce')

In [255]:
events_df['event_date'] = events_df['timestamp'].dt.date

In [256]:
events_df['event_time'] = events_df['timestamp'].dt.time

In [257]:
events_df.dtypes

,0
event_id,int64
session_id,int64
timestamp,datetime64[ns]
event_type,object
product_id,float64
qty,float64
cart_size,float64
payment,object
discount_pct,float64
amount_usd,float64


In [258]:
events_df[['timestamp', 'event_date', 'event_time']].head()

,timestamp,event_date,event_time
0,2021-12-27 00:08:36,2021-12-27,00:08:36
1,2021-12-27 00:16:36,2021-12-27,00:16:36
2,2021-12-27 00:18:01,2021-12-27,00:18:01
3,2021-12-27 00:45:36,2021-12-27,00:45:36
4,2021-12-27 01:03:36,2021-12-27,01:03:36


In [259]:
events_df = events_df.drop(columns=['timestamp'])

In [260]:
events_df.head()

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,event_time
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:08:36
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:16:36
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,00:18:01
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:45:36
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,01:03:36


In [261]:
print (events_df.count () )

event_id        760958
session_id      760958
event_type      760958
product_id      682469
qty             143126
cart_size        44909
payment          33580
discount_pct     33580
amount_usd       33580
event_date      760958
event_time      760958
dtype: int64


In [262]:
events_df.describe()

,event_id,session_id,product_id,qty,cart_size,discount_pct,amount_usd
count,760958.000000,760958.000000,682469.000000,143126.000000,44909.000000,33580.000000,33580.000000
mean,380479.500000,60153.502012,652.313074,1.301748,2.286958,7.160214,133.806357
std,219669.797409,34613.117088,340.279129,0.783436,1.612397,7.474901,152.130611
min,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,2.800000
25%,190240.250000,30314.250000,368.000000,1.000000,1.000000,0.000000,40.300000
50%,380479.500000,60176.000000,674.000000,1.000000,2.000000,5.000000,86.460000
75%,570718.750000,90110.000000,954.000000,1.000000,3.000000,15.000000,174.270000
max,760958.000000,120000.000000,1197.000000,4.000000,13.000000,20.000000,2984.580000


In [263]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
events_df.isna().sum()

Number of NaN values for each columns: 


,0
event_id,0
session_id,0
event_type,0
product_id,78489
qty,617832
cart_size,716049
payment,727378
discount_pct,727378
amount_usd,727378
event_date,0


*NaN values are retained because they are expected based on different customer event types (e.g., page_view, add_to_cart, purchase).*

In [264]:
#Check duplicate
events_df.duplicated().sum()

np.int64(0)

Validate categorical column in event_df

In [265]:
events_df['event_type'].value_counts()

,count
event_type,
page_view,539343
add_to_cart,143126
checkout,44909
purchase,33580


# Analyse orderItems Dataset

In [266]:
orderItems_df.head(10)

,order_id,product_id,unit_price_usd,quantity,line_total_usd
0,1,226,107.15,1,107.15
1,2,771,116.17,1,116.17
2,3,415,94.49,1,94.49
3,3,24,42.86,1,42.86
4,4,1157,32.18,1,32.18
5,5,846,39.60,1,39.60
6,5,745,198.49,1,198.49
7,6,123,81.79,1,81.79
8,6,888,47.85,1,47.85
9,7,405,5.63,1,5.63


In [267]:
orderItems_df.shape

(59163, 5)

In [268]:
orderItems_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59163 entries, 0 to 59162
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        59163 non-null  int64  
 1   product_id      59163 non-null  int64  
 2   unit_price_usd  59163 non-null  float64
 3   quantity        59163 non-null  int64  
 4   line_total_usd  59163 non-null  float64
dtypes: float64(2), int64(3)
memory usage: 2.3 MB


In [269]:
print(orderItems_df.columns.values)

['order_id' 'product_id' 'unit_price_usd' 'quantity' 'line_total_usd']


In [270]:
#showing the datatypes of the column
orderItems_df.dtypes

,0
order_id,int64
product_id,int64
unit_price_usd,float64
quantity,int64
line_total_usd,float64


In [271]:
print(orderItems_df.count())

order_id          59163
product_id        59163
unit_price_usd    59163
quantity          59163
line_total_usd    59163
dtype: int64


In [272]:
#describe the dataframe with showing the count, mean, standard deviation, min, percentiles, and max.
orderItems_df.describe()

,order_id,product_id,unit_price_usd,quantity,line_total_usd
count,59163.000000,59163.000000,59163.000000,59163.000000,59163.000000
mean,16811.777868,692.100147,62.837561,1.303281,81.733655
std,9687.818391,329.197199,62.061297,0.786787,106.360939
min,1.000000,1.000000,3.500000,1.000000,3.500000
25%,8454.000000,411.000000,21.640000,1.000000,24.770000
50%,16838.000000,738.000000,44.650000,1.000000,49.920000
75%,25203.000000,980.000000,79.450000,1.000000,98.680000
max,33580.000000,1197.000000,596.620000,4.000000,2376.920000


In [273]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
orderItems_df.isna().sum()

Number of NaN values for each columns: 


,0
order_id,0
product_id,0
unit_price_usd,0
quantity,0
line_total_usd,0


In [274]:
#Check duplicate
orderItems_df.duplicated().sum()

np.int64(73)

In [275]:
#drop duplicates
orderItems_df= orderItems_df.drop_duplicates()

In [276]:
#Reheck duplicate
orderItems_df.duplicated().sum()

np.int64(0)

In [277]:
orderItems_df.head()

,order_id,product_id,unit_price_usd,quantity,line_total_usd
0,1,226,107.15,1,107.15
1,2,771,116.17,1,116.17
2,3,415,94.49,1,94.49
3,3,24,42.86,1,42.86
4,4,1157,32.18,1,32.18


# Analyse orders Dataset

In [278]:
orders_df.head(10)

,order_id,customer_id,order_time,payment_method,discount_pct,subtotal_usd,total_usd,country,device,source
0,1,13917,2025-01-31T23:07:42,card,20,107.15,85.72,PL,desktop,organic
1,2,1022,2024-02-19T01:17:50,card,0,116.17,116.17,FR,tablet,organic
2,3,6145,2024-12-04T20:24:13,card,0,137.35,137.35,US,mobile,organic
3,4,3152,2024-07-17T08:50:47,card,15,32.18,27.35,BR,mobile,email
4,5,12378,2020-08-21T16:54:16,card,0,238.09,238.09,NL,desktop,paid
5,6,18712,2022-03-07T03:42:34,paypal,0,129.64,129.64,GB,mobile,paid
6,7,17224,2020-03-15T08:56:46,wallet,15,145.45,123.63,JP,mobile,paid
7,8,3448,2021-12-06T01:26:52,card,0,9.03,9.03,JP,desktop,direct
8,9,9132,2022-11-28T03:34:04,card,20,392.95,314.36,IN,mobile,direct
9,10,14764,2023-03-15T10:06:17,paypal,20,9.69,7.75,ES,mobile,organic


In [279]:
orders_df.shape

(33580, 10)

In [280]:
print(orders_df.columns.values)

['order_id' 'customer_id' 'order_time' 'payment_method' 'discount_pct'
 'subtotal_usd' 'total_usd' 'country' 'device' 'source']


In [281]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33580 entries, 0 to 33579
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        33580 non-null  int64  
 1   customer_id     33580 non-null  int64  
 2   order_time      33580 non-null  object 
 3   payment_method  33580 non-null  object 
 4   discount_pct    33580 non-null  int64  
 5   subtotal_usd    33580 non-null  float64
 6   total_usd       33580 non-null  float64
 7   country         33580 non-null  object 
 8   device          33580 non-null  object 
 9   source          33580 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 2.6+ MB


In [282]:
# saparate [order_time] to date and time

orders_df['order_time'] = pd.to_datetime(orders_df['order_time'] )


In [283]:
orders_df['order_date'] = orders_df['order_time'].dt.date

In [284]:
orders_df['order_time'] = orders_df['order_time'].dt.time

In [285]:
orders_df[['order_date', 'order_time']].head()

,order_date,order_time
0,2025-01-31,23:07:42
1,2024-02-19,01:17:50
2,2024-12-04,20:24:13
3,2024-07-17,08:50:47
4,2020-08-21,16:54:16


In [286]:
events_df.columns

Index(['event_id', 'session_id', 'event_type', 'product_id', 'qty',
       'cart_size', 'payment', 'discount_pct', 'amount_usd', 'event_date',
       'event_time'],
      dtype='object')

In [287]:
orders_df.head()

,order_id,customer_id,order_time,payment_method,discount_pct,subtotal_usd,total_usd,country,device,source,order_date
0,1,13917,23:07:42,card,20,107.15,85.72,PL,desktop,organic,2025-01-31
1,2,1022,01:17:50,card,0,116.17,116.17,FR,tablet,organic,2024-02-19
2,3,6145,20:24:13,card,0,137.35,137.35,US,mobile,organic,2024-12-04
3,4,3152,08:50:47,card,15,32.18,27.35,BR,mobile,email,2024-07-17
4,5,12378,16:54:16,card,0,238.09,238.09,NL,desktop,paid,2020-08-21


In [288]:
orders_df = orders_df[[
    'order_id', 'customer_id', 'order_time', 'order_date',
    'payment_method', 'discount_pct', 'subtotal_usd',
    'total_usd', 'country', 'device', 'source'
]]

In [289]:
print(orders_df.columns.values)

['order_id' 'customer_id' 'order_time' 'order_date' 'payment_method'
 'discount_pct' 'subtotal_usd' 'total_usd' 'country' 'device' 'source']


In [290]:

#describe the dataframe with showing the count, mean, standard deviation, min, percentiles, and max.
orders_df.describe()

,order_id,customer_id,discount_pct,subtotal_usd,total_usd
count,33580.000000,33580.000000,33580.000000,33580.000000,33580.000000
mean,16790.500000,10010.629899,7.160214,144.002627,133.806357
std,9693.855356,5798.458913,7.474901,162.603585,152.130611
min,1.000000,1.000000,0.000000,3.500000,2.800000
25%,8395.750000,4964.000000,0.000000,44.060000,40.300000
50%,16790.500000,10008.000000,5.000000,93.620000,86.460000
75%,25185.250000,15061.000000,15.000000,187.100000,174.270000
max,33580.000000,19999.000000,20.000000,3141.660000,2984.580000


Check Null Value and duplicate

In [291]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
orders_df.isna().sum()

Number of NaN values for each columns: 


,0
order_id,0
customer_id,0
order_time,0
order_date,0
payment_method,0
discount_pct,0
subtotal_usd,0
total_usd,0
country,0
device,0


In [292]:
#Check duplicate
orders_df.duplicated().sum()

np.int64(0)

Validate categorical column in customers_df

In [293]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33580 entries, 0 to 33579
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        33580 non-null  int64  
 1   customer_id     33580 non-null  int64  
 2   order_time      33580 non-null  object 
 3   order_date      33580 non-null  object 
 4   payment_method  33580 non-null  object 
 5   discount_pct    33580 non-null  int64  
 6   subtotal_usd    33580 non-null  float64
 7   total_usd       33580 non-null  float64
 8   country         33580 non-null  object 
 9   device          33580 non-null  object 
 10  source          33580 non-null  object 
dtypes: float64(2), int64(3), object(6)
memory usage: 2.8+ MB


In [294]:
orders_df['payment_method'].value_counts()

,count
payment_method,
card,23455
paypal,5032
wallet,3373
cod,1720


In [295]:
orders_df['device'].value_counts()

,count
device,
mobile,18469
desktop,12750
tablet,2361


In [296]:
orders_df['source'].value_counts()

,count
source,
organic,11268
direct,8387
paid,4121
social,4024
email,3056
referral,2724


In [297]:
#Check total number of country
orders_df['country'].nunique()

17

total country same as pervious datasets

In [298]:
orders_df['country'].value_counts()

,count
country,
US,6138
GB,2653
IN,2642
BR,2464
DE,2321
FR,2188
MX,2022
AU,1797
JP,1655


country name in shortforms

In [299]:
#change country name to its full name
#based on previous country_map
print(country_map)

{'AE': 'United Arab Emirates', 'AU': 'Australia', 'BR': 'Brazil', 'CA': 'Canada', 'DE': 'Germany', 'ES': 'Spain', 'FR': 'France', 'GB': 'United Kingdom', 'IN': 'India', 'JP': 'Japan', 'MX': 'Mexico', 'NL': 'Netherlands', 'PL': 'Poland', 'SE': 'Sweden', 'SG': 'Singapore', 'US': 'United States', 'ZA': 'South Africa'}


In [300]:
orders_df['country'] = orders_df['country'].map(country_map)

In [301]:
orders_df['country'].value_counts()

,count
country,
United States,6138
United Kingdom,2653
India,2642
Brazil,2464
Germany,2321
France,2188
Mexico,2022
Australia,1797
Japan,1655


In [302]:
orders_df.head()

,order_id,customer_id,order_time,order_date,payment_method,discount_pct,subtotal_usd,total_usd,country,device,source
0,1,13917,23:07:42,2025-01-31,card,20,107.15,85.72,Poland,desktop,organic
1,2,1022,01:17:50,2024-02-19,card,0,116.17,116.17,France,tablet,organic
2,3,6145,20:24:13,2024-12-04,card,0,137.35,137.35,United States,mobile,organic
3,4,3152,08:50:47,2024-07-17,card,15,32.18,27.35,Brazil,mobile,email
4,5,12378,16:54:16,2020-08-21,card,0,238.09,238.09,Netherlands,desktop,paid


# Analyse products Dataset

In [303]:
products_df.head(10)

,product_id,category,name,price_usd,cost_usd,margin_usd
0,1,Electronics,SSD MediumBlue 149,570.28,352.69,217.59
1,2,Electronics,Keyboard DeepPink 696,498.13,263.13,235.00
2,3,Electronics,Headphones Orchid 188,548.53,309.60,238.93
3,4,Electronics,Smartwatch BurlyWood 664,268.36,153.56,114.80
4,5,Electronics,Smartwatch Cornsilk 328,63.69,42.65,21.04
5,6,Electronics,Headphones WhiteSmoke 326,358.95,222.19,136.76
6,7,Electronics,Headphones DarkGray 529,347.30,187.53,159.77
7,8,Electronics,Mouse Moccasin 673,338.19,227.02,111.17
8,9,Electronics,Mouse Chartreuse 292,88.74,59.58,29.16
9,10,Electronics,SSD Tan 677,85.54,61.04,24.50


In [304]:
products_df.shape

(1197, 6)

In [305]:
print(products_df.columns.values)

['product_id' 'category' 'name' 'price_usd' 'cost_usd' 'margin_usd']


In [306]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   product_id  1197 non-null   int64  
 1   category    1197 non-null   object 
 2   name        1197 non-null   object 
 3   price_usd   1197 non-null   float64
 4   cost_usd    1197 non-null   float64
 5   margin_usd  1197 non-null   float64
dtypes: float64(3), int64(1), object(2)
memory usage: 56.2+ KB


In [307]:
print(products_df.count())

product_id    1197
category      1197
name          1197
price_usd     1197
cost_usd      1197
margin_usd    1197
dtype: int64


In [308]:
products_df.describe()

,product_id,price_usd,cost_usd,margin_usd
count,1197.000000,1197.000000,1197.000000,1197.000000
mean,599.000000,119.935171,77.626483,42.308688
std,345.688444,119.681742,79.228858,44.464057
min,1.000000,3.500000,2.210000,1.030000
25%,300.000000,38.540000,23.530000,12.240000
50%,599.000000,77.200000,49.140000,27.020000
75%,898.000000,165.420000,103.670000,55.440000
max,1197.000000,596.620000,446.100000,263.730000


Check null value and duplicates

In [309]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
products_df.isna().sum()

Number of NaN values for each columns: 


,0
product_id,0
category,0
name,0
price_usd,0
cost_usd,0
margin_usd,0


In [310]:
#Check duplicate
products_df.duplicated().sum()

np.int64(0)

Validate categorical column

In [311]:
products_df['category'].value_counts()

,count
category,
Electronics,171
Home & Kitchen,171
Beauty,171
Sports,171
Fashion,171
Books,171
Toys,171


In [312]:
products_df['name'].nunique()

1197

the product name is unique

In [313]:
#in this project we just want to focus on Electronic category
products_df = products_df.loc[products_df['category'] == 'Electronics']

In [314]:
products_df['category'].value_counts()

,count
category,
Electronics,171


In [315]:
products_df.head()

,product_id,category,name,price_usd,cost_usd,margin_usd
0,1,Electronics,SSD MediumBlue 149,570.28,352.69,217.59
1,2,Electronics,Keyboard DeepPink 696,498.13,263.13,235.00
2,3,Electronics,Headphones Orchid 188,548.53,309.60,238.93
3,4,Electronics,Smartwatch BurlyWood 664,268.36,153.56,114.80
4,5,Electronics,Smartwatch Cornsilk 328,63.69,42.65,21.04


# Analyse reviews Dataset

In [316]:
reviews_df.head(10)

,review_id,order_id,product_id,rating,review_text,review_time
0,1,4,1157,2,Not great quality.,2022-02-01T00:00:00
1,2,7,405,4,Good value for money!,2020-03-07T00:00:00
2,3,7,487,5,"Excellent product, highly recommend!",2024-12-26T00:00:00
3,4,7,442,5,"Excellent product, highly recommend!",2021-03-18T00:00:00
4,5,7,348,4,Good value for money!,2025-09-30T00:00:00
5,6,23,1018,4,Good value for money!,2021-12-11T00:00:00
6,7,33,769,2,Not great quality.,2021-06-04T00:00:00
7,8,36,602,4,Good value for money!,2025-07-10T00:00:00
8,9,36,351,5,"Excellent product, highly recommend!",2020-05-08T00:00:00
9,10,49,501,3,Okay overall.,2023-02-12T00:00:00


In [317]:
reviews_df.shape

(10780, 6)

In [318]:
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10780 entries, 0 to 10779
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   review_id    10780 non-null  int64 
 1   order_id     10780 non-null  int64 
 2   product_id   10780 non-null  int64 
 3   rating       10780 non-null  int64 
 4   review_text  10780 non-null  object
 5   review_time  10780 non-null  object
dtypes: int64(4), object(2)
memory usage: 505.4+ KB


In [319]:
reviews_df['review_time'] = pd.to_datetime(reviews_df['review_time'])


In [320]:
reviews_df['review_date'] = reviews_df['review_time'].dt.date

In [321]:
reviews_df['review_time'] = reviews_df['review_time'].dt.time

In [322]:
reviews_df[['review_date', 'review_time']].head()

,review_date,review_time
0,2022-02-01,00:00:00
1,2020-03-07,00:00:00
2,2024-12-26,00:00:00
3,2021-03-18,00:00:00
4,2025-09-30,00:00:00


In [323]:
reviews_df.head(10)

,review_id,order_id,product_id,rating,review_text,review_time,review_date
0,1,4,1157,2,Not great quality.,00:00:00,2022-02-01
1,2,7,405,4,Good value for money!,00:00:00,2020-03-07
2,3,7,487,5,"Excellent product, highly recommend!",00:00:00,2024-12-26
3,4,7,442,5,"Excellent product, highly recommend!",00:00:00,2021-03-18
4,5,7,348,4,Good value for money!,00:00:00,2025-09-30
5,6,23,1018,4,Good value for money!,00:00:00,2021-12-11
6,7,33,769,2,Not great quality.,00:00:00,2021-06-04
7,8,36,602,4,Good value for money!,00:00:00,2025-07-10
8,9,36,351,5,"Excellent product, highly recommend!",00:00:00,2020-05-08
9,10,49,501,3,Okay overall.,00:00:00,2023-02-12


In [324]:
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10780 entries, 0 to 10779
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   review_id    10780 non-null  int64 
 1   order_id     10780 non-null  int64 
 2   product_id   10780 non-null  int64 
 3   rating       10780 non-null  int64 
 4   review_text  10780 non-null  object
 5   review_time  10780 non-null  object
 6   review_date  10780 non-null  object
dtypes: int64(4), object(3)
memory usage: 589.7+ KB


In [325]:
reviews_df['review_date'] = pd.to_datetime(reviews_df['review_date'])

In [326]:
reviews_df.dtypes

,0
review_id,int64
order_id,int64
product_id,int64
rating,int64
review_text,object
review_time,object
review_date,datetime64[ns]


Check Null Value and duplicates

In [327]:
#shows the number of null values in the dataframe
print("Number of NaN values for each columns: ")
reviews_df.isna().sum()

Number of NaN values for each columns: 


,0
review_id,0
order_id,0
product_id,0
rating,0
review_text,0
review_time,0
review_date,0


In [328]:
#Check duplicate
reviews_df.duplicated().sum()

np.int64(0)

**Validate categorical dataframe**

In [329]:
reviews_df['rating'].value_counts()

,count
rating,
5,4055
4,3565
3,1980
2,757
1,423


the rating for reviews is 1-5

# Hackathon Question


- Analyze User Behavior & Transactions
  - Process user interactions, browsing patterns, and purchase history in real-time
- Predict Cart Abandonment
  - Identify users likely to drop off using behavioral patterns and predictive models
- Recommend Next Best Products
  - Suggest personalized products based on user activity and preferences

# Analyze User Behavior & Transactions

### **Merge Dataframe**

In [330]:
behavior_df = (
    events_df
    .merge(session_df, on="session_id", how="left", suffixes=("_event", "_session"))
    .merge(customers_df, on="customer_id", how="left")
    .merge(products_df, on="product_id", how="left")
)


In [331]:
behavior_df.head()

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,country_y,age,age_group,signup_date,marketing_opt_in,category,name_y,price_usd,cost_usd,margin_usd
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,43,Mid Age,2024-08-12,Unsubscribed,Electronics,Headphones Snow 402,249.08,163.2,85.88
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,43,Mid Age,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,43,Mid Age,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,43,Mid Age,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,43,Mid Age,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN


In [332]:
order_summary = (
    orders_df.groupby("customer_id", as_index=False)
    .agg(
        total_orders=("order_id", "nunique"),
        total_spent=("total_usd", "sum"),
        last_order_time=("order_time", "max")
    )
)

In [333]:
order_summary

,customer_id,total_orders,total_spent,last_order_time
0,1,2,115.39,16:02:53
1,2,2,68.52,17:48:30
2,3,1,66.72,07:39:11
3,4,2,279.86,03:07:16
4,5,3,271.29,21:36:57
...,...,...,...,...
16263,19993,3,670.67,11:44:48
16264,19994,1,263.30,16:48:12
16265,19996,1,65.53,18:42:19
16266,19998,4,568.55,23:42:25


In [334]:
behavior_transactions_df = behavior_df.merge(order_summary, on="customer_id", how="left")

In [335]:
behavior_transactions_df.head(20)

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,signup_date,marketing_opt_in,category,name_y,price_usd,cost_usd,margin_usd,total_orders,total_spent,last_order_time
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,Electronics,Headphones Snow 402,249.08,163.2,85.88,3.0,619.23,14:23:50
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
5,6,1,add_to_cart,946.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
6,7,1,page_view,915.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
7,8,1,page_view,931.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
8,9,1,add_to_cart,931.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
9,10,1,page_view,978.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50


In [336]:
# Validate key relationship
behavior_transactions_df.groupby('session_id')['customer_id'].nunique().max()

1

1 session = 1 user

In [337]:
#Check missing value
behavior_transactions_df.isnull().sum()

,0
event_id,0
session_id,0
event_type,0
product_id,78489
qty,617832
cart_size,716049
payment,727378
discount_pct,727378
amount_usd,727378
event_date,0


*Missing* values were expected due to the event-driven nature of the dataset, where attributes such as quantity and transaction value are only applicable to specific event types.

In [338]:
# Check duplicate
behavior_transactions_df.duplicated().sum()

np.int64(0)

no duplicates in this merge file

In [339]:
behavior_transactions_df['session_id'].duplicated().sum()

np.int64(640958)

1 session → many events

In [340]:
behavior_transactions_df['session_id'].value_counts().head()

,count
session_id,
8287,17
84620,17
119546,17
81630,17
9118,17


In [341]:
behavior_transactions_df[['session_id','customer_id']].head()

,session_id,customer_id
0,1,12360
1,1,12360
2,1,12360
3,1,12360
4,1,12360


In [342]:
behavior_transactions_df.columns

Index(['event_id', 'session_id', 'event_type', 'product_id', 'qty',
       'cart_size', 'payment', 'discount_pct', 'amount_usd', 'event_date',
       'event_time', 'customer_id', 'device', 'source', 'country_x',
       'session_date', 'session_time', 'name_x', 'email', 'country_y', 'age',
       'age_group', 'signup_date', 'marketing_opt_in', 'category', 'name_y',
       'price_usd', 'cost_usd', 'margin_usd', 'total_orders', 'total_spent',
       'last_order_time'],
      dtype='object')

# Check column issues (_x, _y)


In [343]:
behavior_transactions_df.columns

Index(['event_id', 'session_id', 'event_type', 'product_id', 'qty',
       'cart_size', 'payment', 'discount_pct', 'amount_usd', 'event_date',
       'event_time', 'customer_id', 'device', 'source', 'country_x',
       'session_date', 'session_time', 'name_x', 'email', 'country_y', 'age',
       'age_group', 'signup_date', 'marketing_opt_in', 'category', 'name_y',
       'price_usd', 'cost_usd', 'margin_usd', 'total_orders', 'total_spent',
       'last_order_time'],
      dtype='object')

In [344]:
# Validate data consistency
behavior_transactions_df[['country_x','country_y']].head()

,country_x,country_y
0,Germany,Germany
1,Germany,Germany
2,Germany,Germany
3,Germany,Germany
4,Germany,Germany


In [345]:
behavior_transactions_df = behavior_transactions_df.drop(columns=['country_y'])
behavior_transactions_df = behavior_transactions_df.rename(columns={'country_x': 'country'})

In [346]:
behavior_transactions_df.head()

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,signup_date,marketing_opt_in,category,name_y,price_usd,cost_usd,margin_usd,total_orders,total_spent,last_order_time
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,Electronics,Headphones Snow 402,249.08,163.2,85.88,3.0,619.23,14:23:50
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50


In [347]:
behavior_transactions_df[['name_x','name_y']].head()

,name_x,name_y
0,Ronald Williams,Headphones Snow 402
1,Ronald Williams,NaN
2,Ronald Williams,NaN
3,Ronald Williams,NaN
4,Ronald Williams,NaN


name_x = Customer name

name_y = Product name

In [348]:
behavior_transactions_df = behavior_transactions_df.rename(columns={
    'name_x': 'customer_name',
    'name_y': 'product_name'
})

In [349]:
behavior_transactions_df.head()

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,signup_date,marketing_opt_in,category,product_name,price_usd,cost_usd,margin_usd,total_orders,total_spent,last_order_time
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,Electronics,Headphones Snow 402,249.08,163.2,85.88,3.0,619.23,14:23:50
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,2024-08-12,Unsubscribed,NaN,NaN,NaN,NaN,NaN,3.0,619.23,14:23:50


In [350]:
# check merge data type
behavior_transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 760958 entries, 0 to 760957
Data columns (total 31 columns):
 #   Column            Non-Null Count   Dtype   
---  ------            --------------   -----   
 0   event_id          760958 non-null  int64   
 1   session_id        760958 non-null  int64   
 2   event_type        760958 non-null  object  
 3   product_id        682469 non-null  float64 
 4   qty               143126 non-null  float64 
 5   cart_size         44909 non-null   float64 
 6   payment           33580 non-null   object  
 7   discount_pct      33580 non-null   float64 
 8   amount_usd        33580 non-null   float64 
 9   event_date        760958 non-null  object  
 10  event_time        760958 non-null  object  
 11  customer_id       760958 non-null  int64   
 12  device            760958 non-null  object  
 13  source            760958 non-null  object  
 14  country           760958 non-null  object  
 15  session_date      760958 non-null  object  
 16  se

**Rearrange Column name**

In [351]:
behavior_transactions_df.columns.values

array(['event_id', 'session_id', 'event_type', 'product_id', 'qty',
       'cart_size', 'payment', 'discount_pct', 'amount_usd', 'event_date',
       'event_time', 'customer_id', 'device', 'source', 'country',
       'session_date', 'session_time', 'customer_name', 'email', 'age',
       'age_group', 'signup_date', 'marketing_opt_in', 'category',
       'product_name', 'price_usd', 'cost_usd', 'margin_usd',
       'total_orders', 'total_spent', 'last_order_time'], dtype=object)

In [352]:
new_order = [
    # Customer
    'customer_id', 'customer_name', 'email', 'age', 'age_group', 'country',

    # Session
    'session_id', 'device', 'source',

    # Event
    'event_id', 'event_type', 'event_date', 'event_time', 'session_date', 'session_time',

    # Product
    'product_id', 'product_name', 'category',

    # Transaction
    'qty', 'cart_size', 'price_usd', 'discount_pct', 'amount_usd', 'payment', 'last_order_time',

    # Metrics
    'cost_usd', 'margin_usd', 'total_orders', 'total_spent',

    # Marketing
    'marketing_opt_in', 'signup_date'
]

behavior_transactions_df = behavior_transactions_df[new_order]

In [353]:
behavior_transactions_df.columns.values

array(['customer_id', 'customer_name', 'email', 'age', 'age_group',
       'country', 'session_id', 'device', 'source', 'event_id',
       'event_type', 'event_date', 'event_time', 'session_date',
       'session_time', 'product_id', 'product_name', 'category', 'qty',
       'cart_size', 'price_usd', 'discount_pct', 'amount_usd', 'payment',
       'last_order_time', 'cost_usd', 'margin_usd', 'total_orders',
       'total_spent', 'marketing_opt_in', 'signup_date'], dtype=object)

In [354]:
# check value distribution (SANITY CHECK)
behavior_transactions_df['device'].value_counts()


,count
device,
mobile,418289
desktop,289062
tablet,53607


In [355]:
behavior_transactions_df['source'].value_counts()

,count
source,
organic,257354
direct,190197
paid,91985
social,91813
email,69102
referral,60507


In [356]:
behavior_transactions_df['event_type'].value_counts()

,count
event_type,
page_view,539343
add_to_cart,143126
checkout,44909
purchase,33580


In [357]:
# check unique count
behavior_transactions_df['session_id'].nunique()


120000

In [358]:
behavior_transactions_df['customer_id'].nunique()

19945

In [359]:
behavior_transactions_df['product_id'].nunique()

1197

In [360]:
behavior_transactions_df.head(10)

,customer_id,customer_name,email,age,age_group,country,session_id,device,source,event_id,...,discount_pct,amount_usd,payment,last_order_time,cost_usd,margin_usd,total_orders,total_spent,marketing_opt_in,signup_date
0,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,1,...,NaN,NaN,NaN,14:23:50,163.2,85.88,3.0,619.23,Unsubscribed,2024-08-12
1,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,2,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
2,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,3,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
3,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,4,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
4,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,5,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
5,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,6,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
6,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,7,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
7,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,8,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
8,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,9,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12
9,12360,Ronald Williams,angelicamelton@example.net,43,Mid Age,Germany,1,mobile,email,10,...,NaN,NaN,NaN,14:23:50,NaN,NaN,3.0,619.23,Unsubscribed,2024-08-12


In [361]:
behavior_transactions_df.tail(10)

,customer_id,customer_name,email,age,age_group,country,session_id,device,source,event_id,...,discount_pct,amount_usd,payment,last_order_time,cost_usd,margin_usd,total_orders,total_spent,marketing_opt_in,signup_date
760948,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760949,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760949,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760950,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760950,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760951,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760951,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760952,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760952,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760953,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760953,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760954,...,NaN,NaN,NaN,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760954,19550,Danny Chavez,waterstracy@example.org,65,Senior,United States,119999,mobile,referral,760955,...,15.0,74.66,card,15:40:40,NaN,NaN,4.0,406.61,Subscribed,2022-12-08
760955,453,Antonio Lee,michael73@example.com,61,Senior,India,120000,tablet,direct,760956,...,NaN,NaN,NaN,17:38:54,NaN,NaN,2.0,145.88,Subscribed,2022-09-23
760956,453,Antonio Lee,michael73@example.com,61,Senior,India,120000,tablet,direct,760957,...,NaN,NaN,NaN,17:38:54,NaN,NaN,2.0,145.88,Subscribed,2022-09-23
760957,453,Antonio Lee,michael73@example.com,61,Senior,India,120000,tablet,direct,760958,...,NaN,NaN,NaN,17:38:54,NaN,NaN,2.0,145.88,Subscribed,2022-09-23


**Analyse User Behaviour merge file**

In [362]:
user_Behaviour = [
    # Customer
    'customer_id', 'age', 'age_group', 'country',

    # Session
    'session_id', 'device', 'source',

    # Event
    'event_id', 'event_type', 'event_date', 'event_time', 'session_date', 'session_time',

    # Product
    'product_id', 'product_name', 'category',

    # Transaction
    'qty', 'cart_size', 'price_usd', 'discount_pct', 'amount_usd', 'payment', 'last_order_time' ,

    # Metrics
    'total_orders', 'total_spent',

    # Marketing
    'marketing_opt_in'
]

user_Behaviour_df = behavior_transactions_df[user_Behaviour]

In [363]:
user_Behaviour_df.head()

,customer_id,age,age_group,country,session_id,device,source,event_id,event_type,event_date,...,qty,cart_size,price_usd,discount_pct,amount_usd,payment,last_order_time,total_orders,total_spent,marketing_opt_in
0,12360,43,Mid Age,Germany,1,mobile,email,1,page_view,2021-12-27,...,NaN,NaN,249.08,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
1,12360,43,Mid Age,Germany,1,mobile,email,2,page_view,2021-12-27,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
2,12360,43,Mid Age,Germany,1,mobile,email,3,add_to_cart,2021-12-27,...,1.0,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
3,12360,43,Mid Age,Germany,1,mobile,email,4,page_view,2021-12-27,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
4,12360,43,Mid Age,Germany,1,mobile,email,5,page_view,2021-12-27,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed


show the behaviour of customer for example 12360

In [364]:
cust_id = 12360

cust_df = user_Behaviour_df[user_Behaviour_df['customer_id'] == cust_id]

sessions = cust_df['session_id'].unique()
session_count = cust_df['session_id'].nunique()

print("Customer ID:", cust_id)
print("Total Sessions:", session_count)
print("Session IDs:", sessions)

Customer ID: 12360
Total Sessions: 8
Session IDs: [     1  19187  20072  21681  46216  76865 103818 118949]


In [365]:
user_Behaviour_df[user_Behaviour_df['session_id'] == 19187]

,customer_id,age,age_group,country,session_id,device,source,event_id,event_type,event_date,...,qty,cart_size,price_usd,discount_pct,amount_usd,payment,last_order_time,total_orders,total_spent,marketing_opt_in
120689,12360,43,Mid Age,Germany,19187,desktop,paid,120690,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120690,12360,43,Mid Age,Germany,19187,desktop,paid,120691,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120691,12360,43,Mid Age,Germany,19187,desktop,paid,120692,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120692,12360,43,Mid Age,Germany,19187,desktop,paid,120693,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120693,12360,43,Mid Age,Germany,19187,desktop,paid,120694,add_to_cart,2021-05-23,...,1.0,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120694,12360,43,Mid Age,Germany,19187,desktop,paid,120695,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120695,12360,43,Mid Age,Germany,19187,desktop,paid,120696,page_view,2021-05-23,...,NaN,NaN,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120696,12360,43,Mid Age,Germany,19187,desktop,paid,120697,checkout,2021-05-23,...,NaN,1.0,NaN,NaN,NaN,NaN,14:23:50,3.0,619.23,Unsubscribed
120697,12360,43,Mid Age,Germany,19187,desktop,paid,120698,purchase,2021-05-23,...,NaN,NaN,NaN,20.0,58.09,card,14:23:50,3.0,619.23,Unsubscribed


*From the example of analysis above a customer can have multiple sessions, and each session consists of multiple events. Each event represents an action taken by the customer, such as page view, add to cart, checkout or purchase. However, not every session completes all steps of the customer journey.*

In [366]:
# Understand User Activity (Basic EDA)
# Track user journey
user_Behaviour_df['event_type'].value_counts()

,count
event_type,
page_view,539343
add_to_cart,143126
checkout,44909
purchase,33580


In [367]:
event_counts = user_Behaviour_df['event_type'].value_counts()

total = event_counts['page_view']

add_to_cart_rate = event_counts['add_to_cart'] / total
checkout_rate = event_counts['checkout'] / total
purchase_rate = event_counts['purchase'] / total

print (f"Total Page View:  {total}")
print(f"Add to Cart Rate: {add_to_cart_rate:.2%}")
print(f"Checkout Rate: {checkout_rate:.2%}")
print(f"Purchase Rate: {purchase_rate:.2%}")

Total Page View:  539343
Add to Cart Rate: 26.54%
Checkout Rate: 8.33%
Purchase Rate: 6.23%


* Most user activities are page views (393k), but only a small portion convert into purchases (24k)
- While 26.54% of users add items to cart, only 6.23% complete a purchase, indicating a significant drop-off throughout the conversion funnel.*

In [368]:
# Engagement Analysis
# Events per session
user_Behaviour_df.groupby('session_id')['event_type'].count()

,event_type
session_id,
1,10
2,8
3,5
4,2
5,6
...,...
119996,8
119997,8
119998,3


In [369]:
user_Behaviour_df.groupby('session_id')['event_type'].count().mean()

np.float64(6.341316666666667)

*On average, users perform around 6 actions per session, suggesting moderate engagement. This indicates that users interact multiple times within a session before leaving or completing a purchase.*

In [370]:
#Sessions per customer
user_Behaviour_df.groupby('customer_id')['session_id'].nunique()

,session_id
customer_id,
1,5
2,3
3,5
4,9
5,9
...,...
19996,2
19997,4
19998,9


In [371]:
user_Behaviour_df.groupby('customer_id')['session_id'].nunique().mean()

np.float64(6.0165455001253445)

*Most customers have multiple sessions, suggesting repeat engagement and continued interaction with the platform over time.*

In [372]:
user_Behaviour_df.groupby('session_id')['event_type'].count().describe()

,event_type
count,120000.000000
mean,6.341317
std,3.424481
min,1.000000
25%,3.000000
50%,6.000000
75%,9.000000
max,17.000000


In [373]:
purchase_sessions = user_Behaviour_df[user_Behaviour_df['event_type']=='purchase']['session_id'].nunique()
total_sessions = user_Behaviour_df['session_id'].nunique()

conversion_rate = purchase_sessions / total_sessions
print(f"Conversion Rate: {conversion_rate:.2%}")

Conversion Rate: 27.98%


# Predict Cart Abandonment

In [374]:
# merge event_df with session_df
session_events = events_df.merge(session_df, on="session_id", how="left")

In [375]:
session_events.head()

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,event_time,customer_id,device,source,country,session_date,session_time
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:08:36,12360,mobile,email,Germany,2021-12-27,00:01:36
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:16:36,12360,mobile,email,Germany,2021-12-27,00:01:36
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,00:18:01,12360,mobile,email,Germany,2021-12-27,00:01:36
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,00:45:36,12360,mobile,email,Germany,2021-12-27,00:01:36
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,01:03:36,12360,mobile,email,Germany,2021-12-27,00:01:36


In [376]:
cart_abdoment = session_events.merge(
    customers_df,
    on="customer_id",
    how="left"
)

In [377]:
cart_abdoment.head(20)

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,country_x,session_date,session_time,name,email,country_y,age,age_group,signup_date,marketing_opt_in
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
5,6,1,add_to_cart,946.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
6,7,1,page_view,915.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
7,8,1,page_view,931.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
8,9,1,add_to_cart,931.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed
9,10,1,page_view,978.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,Germany,43,Mid Age,2024-08-12,Unsubscribed


In [378]:
cart_abdoment = cart_abdoment.drop(columns=['country_y'])
cart_abdoment = cart_abdoment.rename(columns={'country_x': 'country'})

In [379]:
cart_abdoment.head(20)

,event_id,session_id,event_type,product_id,qty,cart_size,payment,discount_pct,amount_usd,event_date,...,source,country,session_date,session_time,name,email,age,age_group,signup_date,marketing_opt_in
0,1,1,page_view,93.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
1,2,1,page_view,1005.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
2,3,1,add_to_cart,1005.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
3,4,1,page_view,918.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
4,5,1,page_view,946.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
5,6,1,add_to_cart,946.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
6,7,1,page_view,915.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
7,8,1,page_view,931.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
8,9,1,add_to_cart,931.0,1.0,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed
9,10,1,page_view,978.0,NaN,NaN,NaN,NaN,NaN,2021-12-27,...,email,Germany,2021-12-27,00:01:36,Ronald Williams,angelicamelton@example.net,43,Mid Age,2024-08-12,Unsubscribed


# Recommend Next Best Products

In [380]:
purchase_history = (
    orderItems_df
    .merge(orders_df[["order_id", "customer_id", "order_time"]], on="order_id", how="left")
    .merge(products_df, on="product_id", how="left")
)

In [381]:
purchase_history.head()

,order_id,product_id,unit_price_usd,quantity,line_total_usd,customer_id,order_time,category,name,price_usd,cost_usd,margin_usd
0,1,226,107.15,1,107.15,13917,23:07:42,NaN,NaN,NaN,NaN,NaN
1,2,771,116.17,1,116.17,1022,01:17:50,NaN,NaN,NaN,NaN,NaN
2,3,415,94.49,1,94.49,6145,20:24:13,NaN,NaN,NaN,NaN,NaN
3,3,24,42.86,1,42.86,6145,20:24:13,Electronics,SSD Orchid 272,42.86,27.37,15.49
4,4,1157,32.18,1,32.18,3152,08:50:47,NaN,NaN,NaN,NaN,NaN


In [382]:
# based on customer review
recommendation_df = purchase_history.merge(
    reviews_df[["order_id", "product_id", "rating"]],
    on=["order_id", "product_id"],
    how="left"
)

In [383]:
recommendation_df

,order_id,product_id,unit_price_usd,quantity,line_total_usd,customer_id,order_time,category,name,price_usd,cost_usd,margin_usd,rating
0,1,226,107.15,1,107.15,13917,23:07:42,NaN,NaN,NaN,NaN,NaN,NaN
1,2,771,116.17,1,116.17,1022,01:17:50,NaN,NaN,NaN,NaN,NaN,NaN
2,3,415,94.49,1,94.49,6145,20:24:13,NaN,NaN,NaN,NaN,NaN,NaN
3,3,24,42.86,1,42.86,6145,20:24:13,Electronics,SSD Orchid 272,42.86,27.37,15.49,NaN
4,4,1157,32.18,1,32.18,3152,08:50:47,NaN,NaN,NaN,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
59104,33578,1151,9.27,1,9.27,15731,07:34:01,NaN,NaN,NaN,NaN,NaN,NaN
59105,33578,1095,77.20,1,77.20,15731,07:34:01,NaN,NaN,NaN,NaN,NaN,NaN
59106,33578,815,131.78,1,131.78,15731,07:34:01,NaN,NaN,NaN,NaN,NaN,NaN
59107,33579,981,22.22,1,22.22,15971,16:11:57,NaN,NaN,NaN,NaN,NaN,NaN


## **Export Datasets**

In [384]:
#behavior_transactions_df.to_excel(
#   r"C:\Users\Admin\Downloads\E-commerce\behavior_transactions.xlsx",
#    index=False )